scripts: size_report: Add support for TF-M and BL2 image size reports (#43408)
zephyrproject-rtos/zephyr: doc/guides/tfm/build.rst

The TF-M build system creates the following executable files:

* tfm_s - TF-M secure firmware
* tfm_ns - a nonsecure app which is discarded in favor of the Zephyr app

> * bl2 - TF-M MCUBoot, if enabled

Reviewer: ```suggestion
* bl2 - TF-M MCUboot, if enabled
```

The TF-M build system creates the following executable files:

* tfm_s - TF-M secure firmware

> * tfm_ns - TF-M non-secure app (only used by regression tests).

Reviewer: Thanks for pointing this out.

If we only use this for integration tests, wouldn't it be good to introduce a `CONFIG_TFM_NS` to control build of the TF-M ns app, similar to `CONFIG_TFM_BL2` ?
Then integration tests can enable building of TF-M ns when needed.

I see no reason for building the TF-M ns everytime if that is not needed.

If agreeing with this, please create an enhancement task for this work, as I believe  such cleanup to be outside scope of this PR.

Author: Yes, that is a good point. I'll create an internal task for this.

There is a different Kconfig that determines which application is flashed, but it is tied to the regression tests in naming.

Author: Follow up: https://github.com/zephyrproject-rtos/zephyr/pull/43599

   The ``TFM_CMAKE_OPTIONS`` is a list so it is possible to append multiple
   options. Also CMake generator expressions are supported, such as
   ``$<1:-DFOO=bar>``

Footprint and Memory Usage
**************************

The build system offers targets to view and analyse RAM and ROM usage in generated images.
The tools run on the final images and give information about size of symbols and code being used in both RAM and ROM.
For more information on these tools look here: :ref:`footprint_tools`

Use the ``tfm_ram_report`` to get the RAM report for the secure firmware (tfm_s).

.. zephyr-app-commands::
    :tool: all
    :app: samples/hello_world
    :board: mps2_an521_ns
    :goals: tfm_ram_report

Use the ``tfm_rom_report`` to get the ROM report for the secure firmware (tfm_s).

.. zephyr-app-commands::
    :tool: all
    :app: samples/hello_world
    :board: mps2_an521_ns
    :goals: tfm_rom_report

> Use the ``bl2_ram_report`` to get the RAM report for mcuboot, if enabled.

Reviewer: just to be very precise that this is only the MCUboot built as part of TF-M.
```suggestion
Use the ``bl2_ram_report`` to get the RAM report for TF-M enabled MCUboot.
```

Author: The " if enabled " part is just to say that this command is not available when TFM_BL2=n.

Do you wish to separate BL2 from MCUboot in the terminology?
I was aligning with the following text:
```
The TF-M build system creates the following executable files:
* tfm_s - the secure firmware
* tfm_ns - a nonsecure app which is discarded in favor of the Zephyr app
* bl2 - mcuboot, if enabled
```

Should we update to use TF-M MCUboot (or TF-M enabled MCUboot)?

Reviewer 2: That's a fair point ... we should perhaps update this reference as well to be more precise and state: `bl2 - TF-M MCUBoot, if enabled`.

Author: @microbuilder Thanks for the feedback, updated accordingly. Also changed to `TF-M secure firmware`

> Use the ``bl2_ram_report`` to get the RAM report for TF-M MCUBoot, if enabled.

Reviewer: Lower case `b` in MCUboot is the official writing:
```suggestion
Use the ``bl2_ram_report`` to get the RAM report for TF-M MCUboot, if enabled.
```

.. zephyr-app-commands::
    :tool: all
    :app: samples/hello_world
    :board: mps2_an521_ns
    :goals: bl2_ram_report

> Use the ``bl2_rom_report`` to get the ROM report for mcuboot, if enabled.

Reviewer: similar here, I think it is important to make it clear to the reader that this is not just regular MCUboot, but TF-M build MCUboot.
```suggestion
Use the ``bl2_rom_report`` to get the ROM report for TF-M enabled MCUboot.
```

> Use the ``bl2_rom_report`` to get the ROM report for TF-M MCUBoot, if enabled.

Reviewer: Lower case `b` in MCUboot is the official writing:
```suggestion
Use the ``bl2_rom_report`` to get the ROM report for TF-M MCUboot, if enabled.
```
